datafusion/optimizer/src/simplify_expressions/simplify_exprs.rs

```diff
@@ -109,18 +127,22 @@ impl SimplifyExpressions {
             simplifier
         };
 
-        let exprs = plan
-            .expressions()
-            .into_iter()
-            .map(|e| {
+        // the output schema of a filter or join is the input schema. Thus they
+        // can't handle aliased expressions
+        let use_alias = !matches!(plan, LogicalPlan::Filter(_) | LogicalPlan::Join(_));
+        plan.map_expressions(|e| {
+            let new_e = if use_alias {
                 // TODO: unify with `rewrite_preserving_name`
                 let original_name = e.name_for_alias()?;
-                let new_e = simplifier.simplify(e)?;
-                new_e.alias_if_changed(original_name)
-            })
-            .collect::<Result<Vec<_>>>()?;
+                simplifier.simplify(e)?.alias_if_changed(original_name)
+            } else {
+                simplifier.simplify(e)
+            }?;
 
-        plan.with_new_exprs(exprs, new_inputs)
```

Comment:
The whole point of this PR is to remove the calls to `expressions` and `new_with_exprs` (which requires cloning all the expressions)